kv: Improve intent deduplication and testing

tbg opened the pull request:
On commit, TxnCoordSender attaches the collected intent spans to EndTransaction.
There are two sources: spans from previous requests, and spans from the current
batch. Previous handling was asymmetrical and the test was actually broken
(i.e. not asserting properly).
Streamlined both.

@dt after this change you won't be seeing that double intent which highlighted the bug in #4521. You already have a smaller repro so doesn't matter, but wanted to give you a heads up.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/5128)

<!-- Reviewable:end -->

tamird commented:
Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[kv/txn_coord_sender.go, line 366 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5128#-KCZ2BfnT79WmXnmpMAT:-KCZ2BfnT79WmXnmpMAU:-226797404) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e322dc367555b11cff5f55bcef21e4575a0bbd2/kv/txn_coord_sender.go#L366)):*
s/caare/care/

---

*[server/intent_test.go, line 45 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5128#-KCZ31eBbWhhMbx9rkqs:-KCZ31eCFARKRovinJlg:-1622662356) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e322dc367555b11cff5f55bcef21e4575a0bbd2/server/intent_test.go#L45)):*
could you assert this?

---

*[server/intent_test.go, line 92 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5128#-KCZ3bFWwHUEisT0wTGF:-KCZ3bFWwHUEisT0wTGG:767038687) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e322dc367555b11cff5f55bcef21e4575a0bbd2/server/intent_test.go#L92)):*
looks like this is here to help with the debugging `Infof` below - could you return this to the original pattern, or did you intend to leave the `Infof`?

---

*[server/intent_test.go, line 107 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5128#-KCZ3njUvK8zYrvWTTT4:-KCZ3njUvK8zYrvWTTT5:1547881680) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e322dc367555b11cff5f55bcef21e4575a0bbd2/server/intent_test.go#L107)):*
how come `==` -> `>=`? also, if `done` has to be atomic, how come `results` doesn't?

---

*[server/intent_test.go, line 127 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5128#-KCZ43L77LXpqCYq7EyC:-KCZ43L77LXpqCYq7EyD:-378341873) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e322dc367555b11cff5f55bcef21e4575a0bbd2/server/intent_test.go#L127)):*
intended?

---

*[server/intent_test.go, line 137 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5128#-KCZ4ItKWNGtIQeJ2SDB:-KCZ4ItKWNGtIQeJ2SDC:-378341873) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e322dc367555b11cff5f55bcef21e4575a0bbd2/server/intent_test.go#L137)):*
intended?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5128)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[kv/txn_coord_sender.go, line 366 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5128#-KCZ2BfnT79WmXnmpMAT:-KCZAaFBfJXCTjFNcShi:-469593337) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e322dc367555b11cff5f55bcef21e4575a0bbd2/kv/txn_coord_sender.go#L366)):*
done.

---

*[server/intent_test.go, line 45 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5128#-KCZ31eBbWhhMbx9rkqs:-KCZ9o30h2uKHVMPSUld:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e322dc367555b11cff5f55bcef21e4575a0bbd2/server/intent_test.go#L45)):*
Done.

---

*[server/intent_test.go, line 92 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5128#-KCZ3bFWwHUEisT0wTGF:-KCZ9sXuBFfCAWRNZICu:-983528164) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e322dc367555b11cff5f55bcef21e4575a0bbd2/server/intent_test.go#L92)):*
Yeah, was going to leave that in. It's a few lines only.

---

*[server/intent_test.go, line 107 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5128#-KCZ3njUvK8zYrvWTTT4:-KCZAF3XyllPkF0dtwgX:-1803398449) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e322dc367555b11cff5f55bcef21e4575a0bbd2/server/intent_test.go#L107)):*
`done` exists only to avoid double-closing. `results` is a map and I want to collect all that I can.

---

*[server/intent_test.go, line 127 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5128#-KCZ43L77LXpqCYq7EyC:-KCZAQUcNbCtTfTZWHCL:678204588) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e322dc367555b11cff5f55bcef21e4575a0bbd2/server/intent_test.go#L127)):*
yes.

---

*[server/intent_test.go, line 137 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5128#-KCZ4ItKWNGtIQeJ2SDB:-KCZARW5UdzmpLhQQo-X:-1067880085) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e322dc367555b11cff5f55bcef21e4575a0bbd2/server/intent_test.go#L137)):*
yes. I've repeatedly added and removed this stuff over the lifetime of the test.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5128)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Review status: 2 of 4 files reviewed at latest revision, 7 unresolved discussions.

---

*[kv/txn_coord_sender.go, line 371 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5128#-KCZB0EpBQjOJCe3L6Au:-KCZB0EpBQjOJCe3L6Av:-1161456430) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e322dc367555b11cff5f55bcef21e4575a0bbd2/kv/txn_coord_sender.go#L371)):*
If we're serializing this through the `tc.Lock()`, can we store a `RangeGroup` on the `TxnCoordSender` to work with for new writes instead of creating one each time?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5128)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 2 of 4 files reviewed at latest revision, 7 unresolved discussions.

---

*[kv/txn_coord_sender.go, line 371 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5128#-KCZB0EpBQjOJCe3L6Au:-KCZJoedo4tFwGfto0x_:-1517653475) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e322dc367555b11cff5f55bcef21e4575a0bbd2/kv/txn_coord_sender.go#L371)):*
We could do that, or not hold the lock in the case where we're creating the RangeGroup. Went for the solution which is most pleasant to read for now.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5128)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Review status: 2 of 4 files reviewed at latest revision, 7 unresolved discussions.

---

*[kv/txn_coord_sender.go, line 371 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5128#-KCZB0EpBQjOJCe3L6Au:-KCZOFw_wm0g14KDfeit:1534594013) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e322dc367555b11cff5f55bcef21e4575a0bbd2/kv/txn_coord_sender.go#L371)):*
SGTM, we can optimize this if it becomes an issue later.

On a related note though, any rough idea of the relative frequencies through this code path between existing writes and new writes? Should we expect one to happen substantially more than the other?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5128)*
<!-- Sent from Reviewable.io -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[server/intent_test.go, line 107 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5128#-KCZ3njUvK8zYrvWTTT4:-KCa2oTSFGVgDa-iSmCs:-1581599395) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e322dc367555b11cff5f55bcef21e4575a0bbd2/server/intent_test.go#L107)):*
I don't think that answers my question - it seems that `results` can be accessed concurrently, why isn't access to it synchronized? If it doesn't need to be synchronized for whatever reason, why is `done` an `atomic.Value` rather than a plain `bool`?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5128#-:-KCa3XsPocmFwUc-_Sak:1417088721)*
<!-- Sent from Reviewable.io -->

tbg committed:
kv: Improve intent deduplication and testing

On commit, TxnCoordSender attaches the collected intent spans to EndTransaction.
There are two sources: spans from previous requests, and spans from the current
batch. Previous handling was asymmetrical and the test was actually broken
(i.e. not asserting properly).
Streamlined both.

tschottdorf commented:
Review status: 3 of 4 files reviewed at latest revision, 2 unresolved discussions.

---

*[kv/txn_coord_sender.go, line 371 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5128#-KCZB0EpBQjOJCe3L6Au:-KCa9vCXIYj5cmuZ0fbK:-435429678) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e322dc367555b11cff5f55bcef21e4575a0bbd2/kv/txn_coord_sender.go#L371)):*
this is hit exclusively for 1PC transactions, so the simple insert benchmark should give you a good idea of the isolated impact of this.

---

*[server/intent_test.go, line 107 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5128#-KCZ3njUvK8zYrvWTTT4:-KCa9gvcZxiZNdE0NWWn:1573096185) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e322dc367555b11cff5f55bcef21e4575a0bbd2/server/intent_test.go#L107)):*
Ah, valid point. The answer is that I was too tired too think about it yesterday and that there's in fact a mutex already guarding the interior of the command filter (and results is only accessed outside when the Stopper is done). Removed the `atomic.Value`.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5128)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5128)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
oh wow, a green build? I must be dreaming.
